# Promise
```js
function a() {
  console.log('A')
}
function b() {
  console.log('B')
}
b() // B 
a() // A

// 동기 방식 (Synchronous)
// 코드가 작성한 순서대로 동작하는 방식

a() // A
b() // B
``` 
```js
function a() {
  setTimeout(() => {
    console.log('A')
  }, 1000)
}
function b() {
  console.log('B')
}
a()
b()
// B
// A
// setTimeout을 사용하였기에, 순서대로 실행되지 않는다
```
기본적인 처리가 지연되는 로직이 있는 경우, 비동기 개념을 도입하여, A가 먼저 출력되도록 할 수 있다

```js
function a(callback) {
  setTimeout(() => {
    console.log('A')
    callback()
  }, 1000)
}
function b() {
  console.log('B')
}
a(function () {
  b()
})
// A
// B
```
특정한 로직이 끝난 다음에, 또 다른 로직이 실행되는 것을 콜백함수를 통해 보장하였다

`callback`함수는 결국 하나의 데이터이다

```js
function a(callback) {
  const str = 'Hello A'
  setTimeout(() => {
    console.log('A')
    callback(str)
  }, 1000)
}
function b() {
  console.log('B')
}
a(function (event) {
  console.log(event)
  b()
})
// A
// Hello A
// B
```
```js
function a(callback) {
  setTimeout(() => {
    console.log('A')
    callback()
  }, 1000)
}
function b(callback) {
  setTimeout(() => {
    console.log('B')
    callback()
  }, 1000)
}
a(function () {
  b(function () {
    console.log('Done!')
  })
})
// A
// B
// Done!
```

## 기본적인 callback 패턴
```js
function getMovie(cb) {
  fetch(`https://www.omdbapi.com/?apikey=qw12&s=frozen`)
    .then(res => res.json())
    .then(res => {
      cb(res)
    })
}

getMovie(movies => {
  console.log(movies)
  비동기1(() => {
    비동기2(() => {
      비동기3(() => {
        비동기4(() => {
        })
      })
    })
  })
})
```
callback hell 이라는 심각한 문제가 존재한다 

## Callback 지옥..
```js
function a(callback) {
  setTimeout(() => {
    console.log('A')
    callback()
  }, 1000)
}
function b(callback) {
  setTimeout(() => {
    console.log('B')
    callback()
  }, 1000)
}
function c(callback) {
  setTimeout(() => {
    console.log('C')
    callback()
  }, 1000)
}
function d(callback) {
  setTimeout(() => {
    console.log('D')
    callback()
  }, 1000)
}
a(function () {
  b(function () {
    c(function () {
      d(function () {
        console.log('Done!')
      })
    })    
  })
})
```

## `promise` 객체
비동기를 통해서 로직을 기다렸다가, 처리가 끝나면 그 처리가 온전하게 완료되거나 실패할 수 있는데, 그것에 대한 결과를 반환한다 

결국, 무엇인가의 로직을 처리가 완료 될 때 까지 기다린다는 것  
`promise` 단어는 약속한다는 의미를 가진다. 기다리고 난 이후의 실행을 약속한다!

```js
function a() {
  // promise: 약속의 객체!
  return new Promise(resolve => {
    setTimeout(() => {
      console.log('A')
      resolve('Hello A')
    }, 1000)
  })
}
function b() {
  console.log('B')
}
async function test() {
  const res = await a()
  console.log('res:', res)
  b()
}
test()
// A
// res: Hello A
// B
```

```js
function a() {
  // promise: 약속의 객체!
  return new Promise(resolve => {
    setTimeout(() => {
      console.log('A')
      resolve('Hello A')
    }, 1000)
  })
}function b() {
  return new Promise(resolve => {
    setTimeout(() => {
      console.log('B')
      resolve('Hello B')
    }, 1000)
  })
}function c() {
  return new Promise(resolve => {
    setTimeout(() => {
      console.log('C')
      resolve('Hello C')
    }, 1000)
  })
}function d() {
  return new Promise(resolve => {
    setTimeout(() => {
      console.log('D')
      resolve('Hello D')
    }, 1000)
  })
}

async function test() {
  await a()
  await b()
  await c()
  await d()
  console.log('Done!')
}
test()
// A
// B
// C
// D
```


## callback hell을 promise로 변환해보자
```js
function getMovie() {
  // promise instance
  return new Promise(resolve => {
    fetch(`https://www.omdbapi.com/?apikey=qw12&s=frozen`)
      .then(res => res.json())
      .then(res => {
        resolve(res)
      })
  }) 
}

// 1번 `then` 패턴
// getMovie가 호출 된 이후에는 Promise instance가 남는다
getMovie().then(res => console.log(res))

// 2번 `await` 패턴
// async await 패턴에서는 return으로 반환된다
;(async function () {
  const res = await getMovie()
  console.log(res)
})()
```

```js
function imageLoad(src, cb) {
  const imgEl = document.createElement('img')
  imgEl.src = src
  imgEl.addEventListener('load', () => {
    cb()
  })
}

const imgContainer = document.querySelector('.image')

imageLoad('', () => {
  console.log('Done!')
  imgContainer.classList.add('done')
})
```
```js
function imageLoad(src) {
  return new Promise(resolve => {
    const imgEl = document.createElement('img')
    imgEl.src = src
    imgEl.addEventListener('load', () => {
      resolve()
    })
  })
}

const imgContainer = document.querySelector('.image')

imageLoad(('')
  .then(() => {
    console.log('Done!')
    imgContainer.classList.add('done')
  })
)

;(async function () {
  await imageLoad('')
  console.log('Done!')
  imgContainer.classList.add('done)
})
```

```js
async function getMovie() {
  let res = await fetch(`https://www.omdbapi.com/?apikey=qw12&s=frozen`)
  res = await res.json()
  // res.json() 의 결과도 promise instance를 반환한다, 때문에 await 키워드를 작성할 수 있다
  console.log(res) // 영화 정보
  return res
}

const res = getMovie()
console.log(res) // Promise instance

const res2 = await getMovie()
console.log(res2 ) // 영화 정보
```

## 비동기의 예외처리 
에러가 발생할 수 있는 구문을 `try` 부분에 작성하고,  
실제로 `error가` 발생하였을 때는 `catch` 구문으로 넘겨서 처리할 수 있도록 만들 수 있다

### `try` / `catch` / `finally`
```js
function a(number) {
  return new Promise((resolve, reject) => {
    if (number > 4) {
      reject()
      return
    }
    setTimeout(() => {
      console.log('A')
      resolve()
    , 1000})
  })
}

async function test() {
  // a(1)
  //   .then(() => {
  //     console.log('Resolve!')
  //   })
  //   .catch(() => {
  //     console.log('Reject!')
  //   })
  //   .finally(() => {
  //     console.log('Done!')
  //   }) 아래와 같은 내용이다
  try {
    await a(8)
    console.log('Resolve!')
  } catch (error) {
    console.log('Reject!')
  } finally {
    console.log('Done')
    // 로직이 정상적으로 처리 되거나, 거부되어도 finally문은 작동한다
  }
}
test()
// Reject!
```

```js
async function getMovie() {
  try { // 에러가 발생한다면, try문 안에서, 에러가 난 코드 밑으로는 실행이 안된다
    let res = await fetch(`https://www.omdbapi.com/?apikey=qw12&s=frozen`)
    res = await res.json()
    console.log(res)
    return res
  } catch (error) { // 에러가 발생했을 때, 실행되는 코드
    alert('에러 발생')
  } finally { // 무조건 실행되는 코드
    console.log('실행 완료')
  }
}
// try 문은 catch 문과 같이 사용해야한다
```

### `.then()` `.catch()` `.finally()`
```js
const res = getMovie()
console.log(res)
```
```js
function getMovie() {
  return fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen')
    .then(res => res.json())
    .then(res => res)
    .catch(error => alert('~')
    .finally(() => { console.log('Done!')})
}

const res = getMovie()
console.log(res)
```

## 실행 순서 보장
```js
function getMovie() {
  fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen') // 이 코드가 출발하고
    .then(res => res.json()) // 얘네는 나중에 작동한다..
    .then(res => {
      console.log('fetch1:', res)
    })
  
  fetch('https://www.omdbapi.com/?apikey=qw12&s=ironman') // 바로 이 코드가 출발한다
    .then(res => res.json())
    .then(res => {
      console.log('fetch2:', res)
    })
}
// 출발은 동시에 한다. 만약 순서를 보장하고 싶다면?

function getMovie() {
  fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen') 
    .then(res => res.json()) 
    .then(res => {
      console.log('fetch1:', res)

      fetch('https://www.omdbapi.com/?apikey=qw12&s=ironman') 
        .then(res => res.json())
        .then(res => {
          console.log('fetch2:', res)
        })
    })
} 

```
## Promise static methods

### Promise.all()
```js
function getMovie() {
  Promise.all([
    fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen') 
      .then(res => res.json()),
    fetch('https://www.omdbapi.com/?apikey=qw12&s=ironman') 
      .then(res => res.json())
  ])
    .then(res) => {  
      console.log('fetch1:', res[0])
      console.log('fetch2:', res[1])
      console.log('Done!')
    })
}
// 실행 순서가 고정이 되었다
// Promise.all()은 한번에 처리한다, 즉 병렬 처리
```
```js
async function getMovie() {
  const res1 = await fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen')
    .then(res => res.json())
  const res2 = await fetch('https://www.omdbapi.com/?apikey=qw12&s=ironman')
    .then(res => res.json())

  console.log('async1:', res1)
  console.log('async2:', res2)
  console.log('Done!')
}
// 해당 코드의 경우, 직렬 처리 방식이므로, 병렬처리에 비해 시간이 오래 걸린다
```
```js
async function getMovie() {
  const [res1, res2] = await Promise.all([
    fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen')
      .then(res => res.json())
    fetch('https://www.omdbapi.com/?apikey=qw12&s=ironman')
      .then(res => res.json())
  ])

  console.log('async1:', res1)
  console.log('async2:', res2)
  console.log('Done!')
}
getMovie()
```

### Promise.resolve()
인수로 넣은 데이터를 가지고 있는, Promise instance를 만드는 static method 이다
```js
const pi = Promise.resolve()
pi.then(res => console.log(res))
// 1234
```

### Promise.reject()
```js
const p1 = Promise.resolve() // 해결이 된 상태의 instance
const p2 = Promise.reject() // 거부가 된 상태의 instance

p1.then() // O
p2.then() // X

p1.catch() // X
p2.catch() // O
```

### reject
```js
function getMovie() {
  return new Promise((resolve, reject) => {
    fetch('https://www.omdbapi.com/?apikey=qw12&s=frozen')
      .then(res => res.json())
      .then(res => {
        resolve(res)
      })
      .catch(() => {
        reject('문제가 생김!')
      })
  })
}

getMovie()
  .then() // 에러 없을 때, 작동 
  .catch(err => {
    console.log(err)
  }) // 문제가 생김!

// resolve와 then
// reject와 catch를 연관지어 생각하자
```

```js
function imageLoad(src) {
  return new Promise((resolve, reject) => {
    if (!src) {
      reject('이미지 경로 없음')
      return
    }
    const imgEl = document.createElement('img')
    imgEl.src = src
    imgEl.addEventListener('load', () => {
      resolve('이미지 로드 완료')
    })
  })
}

imageLoad()
  .then(res => {
    console.log(res)
  })
  .catch(error => {
    console.log(error)
  })
```

## 대기(pending), 이행(fulfilled) ,거부(rejected)
```js
function fetchMovies(title) {
  // 대기(pending): 이행하거나 거부되지 않은 초기 상태
  const OMDB_API_KEY = 'qw12'
  return new Promise(async (resolve, reject) => {
    try {
      const res = await axios.get(`https://www.omdbapi.com/?apikey=${OMDB_API_KEY}&s=${title}`)
      // 이행(fulfilled): 연산이 성공적으로 완료됨.
      resolve(res)
    } catch ({ message: msg }) {
      console.log(msg)
      // 거부(rejected): 연산이 실패함.
      reject('Hee??')
    }
  })
}

async function test() {
  try {
    const res = await fetchMovies('frozen')
    console.log(res)
  } catch(hee) {
    console.log(hee)
  }
}
test()

function hello() {
  fetchMovies('jobs')
    .then(res => console.log(res))
    .catch(() => { console.log(hee) })
}
hello()

```